apis/apim/v1alpha1/ratelimitpolicy_types.go

```diff
@@ -39,10 +39,14 @@ type GenericKeySpec struct {
 	DescriptorKey *string `json:"descriptor_key,omitempty"`
 }
 
-type MetadataPathSegment struct {
+type MetadataPathSegmentKey struct {
 	Key string `json:"key"`
 }
 
+type MetadataPathSegment struct {
+	Segment MetadataPathSegmentKey `json:"segment"`
+}
+
```

Comment:
```suggestion
type MetadataPathSegmentKey string `json:"key"`

type MetadataPathSegment struct {
        Key MetadataPathSegmentKey `json:"key"`
}

```

or simply:

```suggestion
type MetadataPathSegment struct {
        Key string `json:"key"`
}

```

Reply:
We must follow the exact `structs` definition as in `wasm-shim`. That simplification might not produce the correct result required, I guess.

Reply:
According to [type.metadata.v3.MetadataKey](https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/metadata/v3/metadata.proto#envoy-v3-api-msg-type-metadata-v3-metadatakey) it is 
```yaml
key: envoy.xxx
path:
- key: prop
- key: foo
```

So maybe we can get rid of that `segment`

Reply:
```rust
pub struct MetadataKey {
    // message fields
    pub key: ::std::string::String,
    pub path: ::protobuf::RepeatedField<MetadataKey_PathSegment>,
    // special fields
    #[cfg_attr(feature = "with-serde", serde(skip))]
    pub unknown_fields: ::protobuf::UnknownFields,
    #[cfg_attr(feature = "with-serde", serde(skip))]
    pub cached_size: ::protobuf::CachedSize,
}


pub struct MetadataKey_PathSegment {
    // message oneof groups
    pub segment: ::std::option::Option<MetadataKey_PathSegment_oneof_segment>,
    // special fields
    #[cfg_attr(feature = "with-serde", serde(skip))]
    pub unknown_fields: ::protobuf::UnknownFields,
    #[cfg_attr(feature = "with-serde", serde(skip))]
    pub cached_size: ::protobuf::CachedSize,
}

pub enum MetadataKey_PathSegment_oneof_segment {
    key(::std::string::String),
}
```

Reply:
@eguzki Does that help answer the question?

Reply:
I am afraid, the `segment` is necessary. I do not understand why, though. 

If you try to parse (using the rust parser):
```json
"metadata": {                                      
    "descriptor_key": "user-id",                   
    "default_value": "no-user",                    
   "metadata_key": {                              
       "key": "envoy.filters.http.ext_authz",     
       "path": [                                  
           {                                      
               "key": "ext_auth_data"             
           }, 
           {                                      
               "key": "user_id"             
           }                                    
       ]                                          
    },                                             
    "source": "DYNAMIC"         
}
```

The parser does not return error. And it should IMO. However, when you read the data
```rust
assert_eq!(metadata_path.len(), 2);                     // OK
assert_eq!(metadata_path[0].get_key(), "ext_auth_data");   // NOT OK
```
you get
```
thread 'configuration::test::parse_config' panicked at 'assertion failed: `(left == right)`
  left: `""`,
 right: `"ext_auth_data"`', src/configuration.rs:185:13
```

Reply:
Yup, that's the reasoning behind that test as well. I saw it when I was running limitador in debug mode and keys aren't present if the segment is not provided.

Reply:
We can, if we want to, remove `segment` part for the users by custom serialization from manifest to wasm config.